package/MDAnalysis/topology/ITPParser.py

```diff
@@ -572,13 +582,7 @@ def parse(self, include_dir='/usr/local/gromacs/share/gromacs/top/',
             if all(vals):
                 attrs.append(Attr(np.array(vals, dtype=dtype)))
 
-        if not all(self.masses):
```

Comment:
Arf! I should have seen this coming.

Here, the parser only guesses *some* of the masses, which the new system does not let you do. What the code does here is to read the values that are in the file and guess the rest.

I am not sure how we can achieve this without breaking backwards compatibility :thinking: 

From what I can see, this is the only place where we have this pattern. Everywhere else is all or nothing. PDB-related parsers set masses to 0 which is an issue in its own right.

I think I need more opinions here...

Reply:
🤔 we definitely can't change this behaviour without warning, it looks like a breaking change.

A maybe naive thought here, could we not just implement a "partial guesser" which fixes partially filled attributes?

In fact, this is something we probably would need for element readers like PDB which set missing elements at '', with the idea that folks would then re-guess if for the missing elements if necessary.

Reply:
Having the ability to guess an attribute only for parts of the atoms would be god indeed. I just do not know how to manage that without breaking compatibility in that specific case.

Reply:
I think the idea of partial guessing an attribute is pretty important in improving the functionality and flexibility of the API.  I'll think of an approach to it and propose it to you in a separate commit

Reply:
I think I have an idea of how to solve this issue.

1. define a placeholder value for TopologyAttributes. This would be a class attribute to each TopologyAttribute of interest; only the mass at first.
2. set the placeholder for the mass to `nan`
3. the ITP parser sets the unavailable masses to the placeholder value
4. by default, the guessers fill in the partial guesses
5. filling in partial guesses means guessing the value when it is set to the placeholder value

With the plan above, we do not change the default behaviour *with this PR*: by default, a universe created with the ITP parser has a mixture of messes read from files or guessed when they are missing. No other parser is affected because the ITP parser is the only one with this behaviour.

If we do not ask to guess the masses, then we only have the masses we read from files. The missing masses are well identified.

This would warrant some changes for 3.0, though:
* on the bright side, it would make it easier to fix #3218. When a mass cannot be guessed, we can set it to the placeholder (here `nan`) rather than 0.
* on the less bright side, we have to touch the PDB parser to clean it up with regards to partial attributes. We also have to go through all the other parsers to maintain some consistency. While the result is very good, it will be a chore.